dbms/src/Storages/DeltaMerge/ColumnFile/ColumnFileSetReader.cpp

```diff
@@ -157,7 +157,17 @@ size_t ColumnFileSetReader::readRows(MutableColumns & output_columns, size_t off
             continue;
 
         auto & column_file_reader = column_file_readers[file_index];
-        actual_read += column_file_reader->readRows(output_columns, rows_start_in_file, rows_in_file_limit, range);
+        auto [read_offset, read_rows] = column_file_reader->readRows(output_columns, rows_start_in_file, rows_in_file_limit, range);
+        actual_read += read_rows;
+        if (row_ids != nullptr)
+        {
+            auto rows_before_cur_file = file_index == 0 ? 0 : column_file_rows_end[file_index - 1];
+            auto start_row_id = read_offset + rows_before_cur_file;
+            for (size_t i = 0; i < read_rows; ++i)
+            {
+                row_ids->push_back(start_row_id + i);
+            }
+        }
```

Comment:
usually, when MVCC, the remaining rows are more than the filter-out rows, maybe we can record the filter-out row_id?